src/Datasource/EntityTrait.php

```diff
@@ -283,6 +320,23 @@ public function &get($property)
             throw new InvalidArgumentException('Cannot get an empty property');
         }
 
+        $nested = false;
+        if (strpos($property, '.') > 0 && !array_key_exists($property, $this->_properties)) {
+            list($property, $nested) = explode('.', $property, 2);
+        }
+
+        if ($nested) {
+            if ($this->_properties[$property] instanceof EntityInterface) {
+                $result = $this->_properties[$property]->get($nested);
+                return $result;
+            } elseif (is_array($this->_properties[$property])
+                || ($this->_properties[$property] instanceof \ArrayAccess)
+            ) {
+                $value = Hash::get($this->_properties[$property], $nested);
+                return $value; //@TODO: Or remove this line and let getters engage below?
```

Comment:
But remember that `b` isn't a direct property on `$entity` in this case. Even if we had an `$entity::_getB()` method, it would apply to `$entity->b` and not to `$entity->a['b']`. In this case, `$entity::_getA()` is the last possible getter method involved since we switch over to `Hash::get()` in order to retrieve `b` from `a`. If `$entity->a` was _also_ an Entity, then the code already descends deeper and returns `$entity->a->_getB()` instead. Although...

I need to refactor and experiment a bit. I'm rethinking an [earlier conversation with @markstory about engaging setters recursively](https://github.com/cakephp/cakephp/pull/7847#issuecomment-168219991). In the case of set(), we should use get() all the way down to the last Entity or array possible before calling ::set() or Hash::set(). But with ::get(), we should be invoking every possible getter along the chain:

```php
// These two are already equivalent (assuming the getter methods are present):
$entity->a->b->c;
$entity->_getA()->_getB()->_getC();

// We want this to **also** be equivalent:
$entity->get('a.b.c');
```